practice/Capstone/main.py

```diff
@@ -0,0 +1,249 @@
+        if gen is None:
+            logger.error('Wrong type of data')
+            sys.exit(1)
+        else:
+            data_structure[key] = gen(right)
+    return data_structure
+
+
+def check_path(path: str) -> bool:
+    if os.getcwd() in os.path.abspath(path):
+        return True
+    return False
+
+
+def create_prefixes(number_of_prefixes: int, prefix: str) -> list:
+    if prefix == 'count':
+        return [str(i) for i in range(number_of_prefixes)]
+    elif prefix == 'random':
+        return random.sample(range(100000), number_of_prefixes)
+    elif prefix == 'uuid':
+        return [str(uuid.uuid4()) for _ in range(number_of_prefixes)]
+    else:
+        logger.error('Prefixes can only have value within [\'count\', \'random\', \'uuid\']')
+        sys.exit(1)
+
+
+def single_data_creator(data_pattern: dict) -> dict:
+    return {key: value() for key, value in data_pattern.items()}
+
+
+def data_creator(data_schema: dict, number_of_lines: int) -> List[Dict]:
+    data_pattern = create_pattern_on_based_schema(data_schema)
+    return [single_data_creator(data_pattern) for _ in range(number_of_lines)]
+
+
+def dump_data_to_json(path: str, data) -> None:
+    with open(path + '.jsonl', 'a') as f:
+        for record in data:
+            f.write(json.dumps(record))
+            f.write('\n')
+
+
+def get_data_schema(args: Args) -> dict:
+    try:
+        data_schema = json.loads(args.data_schema)
+    except json.decoder.JSONDecodeError:
+        try:
+            with open(args.data_schema, 'r') as f:
+                data_schema = json.load(f)
+        except FileNotFoundError:
+            logger.error(f'Path {args.data_schema} is not correct')
+            sys.exit(1)
+    return data_schema
+
+
+def clear_path(path_to_save_files: str, file_name: str) -> None:
+    [os.remove(path_to_save_files + '/' + file) for file in os.listdir(path_to_save_files) if
+     file.startswith(file_name)]
+
+
+def output_manager(args: Args) -> None:
+    data_schema = get_data_schema(args)
+    if args.files_count == 0:
+        logger.info('Creating data...')
+        data = data_creator(data_schema, args.data_lines)
+        logger.info('Printing output...')
+        [print(json.dumps(data[i])) for i in range(len(data))]
+    else:
+        if not check_path(args.path_to_save_files):
+            logger.error('Path need to be in directory.')
+            sys.exit(1)
+        if not os.path.exists(args.path_to_save_files):
+            logger.info('Creating output directory...')
+            os.makedirs(args.path_to_save_files)
+        if args.clear_path:
+            logger.info('Clearing path...')
+            clear_path(args.path_to_save_files, args.file_name)
+        logger.info('Creating data...')
+        with multiprocessing.Pool(processes=min(os.cpu_count(), args.multiprocessing)) as pool:
+            data = pool.starmap(partial(data_creator, data_schema=data_schema, number_of_lines=args.data_lines),
```

Comment:
could you explain what happens in your .starmap() statement and why did you choose this method?

Reply:
Sure, we are invoking args.files_count times function data_creator(data_schema, args.data_lines). I chose this method, because we want to invoke data_creator function with the same arguments all the time and I'm not sure if it is possible to do it with .map().

Reply:
Okay, thanks, good explanation. Map() is also possible because you have already created functions with predefined parameters via partial()